src/lib/HST/Environment.hs

```diff
@@ -1,101 +1,70 @@
+    Environment(..)
     -- * Lookup
-  , lookupConEntry
-  , lookupDataEntry
-    -- * Insertion
-  , insertConEntry
-  , insertDataEntry
+  , lookupConEntries
+  , lookupTypeName
   ) where
 
-import           Data.Map.Strict     ( Map )
-import qualified Data.Map.Strict     as Map
-
-import qualified HST.Frontend.Syntax as S
-
--------------------------------------------------------------------------------
--- Aliases for Names in the Environment                                      --
--------------------------------------------------------------------------------
--- | The name of a data type.
-type TypeName a = S.QName a
-
--- | The name of a data constructor.
-type ConName a = S.QName a
-
--------------------------------------------------------------------------------
--- Environment Entries                                                       --
--------------------------------------------------------------------------------
--- | An entry of the 'Environment' for a data constructor that is in scope.
-data ConEntry a = ConEntry
-  { conEntryName    :: ConName a
-    -- ^ The name of the constructor.
-  , conEntryArity   :: Int
-    -- ^ The number of fields of the constructor.
-  , conEntryIsInfix :: Bool
-    -- ^ Whether the constructor should be written in infix notation or not.
-  , conEntryType    :: TypeName a
-    -- ^ The name of the data type that the constructor belongs to.
-  }
+import qualified Data.Map.Strict        as Map
+import           Data.Maybe             ( mapMaybe )
 
--- | An entry of the 'Environment' for a data type whose constructors are in
---   scope.
-data DataEntry a = DataEntry
-  { dataEntryName :: TypeName a
-    -- ^ The name of the data type.
-  , dataEntryCons :: [ConName a]
-    -- ^ The names of the data type's constructors.
-  }
+import           HST.Effect.InputModule
+  ( ConEntry, ConName, ModuleInterface(..), TypeName )
+import qualified HST.Frontend.Syntax    as S
 
 -------------------------------------------------------------------------------
 -- Environment                                                               --
 -------------------------------------------------------------------------------
--- | A data type for the state of the pattern matching compiler.
+-- | A data type for the environment of the pattern matching compiler
+--   containing data types and data constructors currently in scope.
 data Environment a = Environment
-  { envConEntries  :: Map (ConName a) (ConEntry a)
-    -- ^ Maps names of constructors to their 'ConEntry's.
-  , envDataEntries :: Map (TypeName a) (DataEntry a)
-    -- ^ Maps names of data types to their 'DataEntry's.
+  { envCurrentModule   :: ModuleInterface a
+    -- ^ The module interface for the current module, containing its data types
+    --   and data constructors.
+  , envImportedModules :: [(S.ImportDecl a, ModuleInterface a)]
+    -- ^ A list of all successfully imported module interfaces including their
+    --   import declarations. 
```

Comment:
Is the `S.ImportDecl a` every used? I could only find `map snd (envImportedModules env)`.

Reply:
It is used in my implementation of #12, which I will create a pull request for soon.